File: crates/types/src/scheduler.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ScheduledTaskConfigValueType {
    Integer,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct ScheduledTaskConfigField {
    pub key: String,
    pub label_key: String,
    pub value_type: ScheduledTaskConfigValueType,
    pub required: bool,
    pub min: Option<i64>,
    pub max: Option<i64>,
    pub unit_key: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct ScheduledTaskDefinition {
    pub code: String,
    pub name_key: String,
    pub description_key: String,
    pub default_enabled: bool,
    pub default_interval_seconds: i64,
    pub default_config: serde_json::Value,
    pub config_schema: Vec<ScheduledTaskConfigField>,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct ScheduledTask {
    pub code: String,
    pub name_key: String,
    pub description_key: String,
    pub enabled: bool,
    pub interval_seconds: i64,
    pub config: serde_json::Value,
    pub config_schema: Vec<ScheduledTaskConfigField>,
    pub last_started_at: Option<String>,
    pub last_finished_at: Option<String>,
    pub last_status: Option<String>,
    pub last_duration_ms: Option<i64>,
    pub last_error: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Clone, Debug, Default, PartialEq, Deserialize)]
pub struct ScheduledTaskUpdate {
    #[serde(default)]
    pub enabled: Option<bool>,
    #[serde(default)]
    pub interval_seconds: Option<i64>,
    #[serde(default)]
    pub config: Option<serde_json::Value>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct ScheduledTaskRun {
    pub id: String,
    pub task_code: String,
    pub status: String,
    pub started_at: String,
    pub finished_at: Option<String>,
    pub duration_ms: Option<i64>,
    pub message: Option<String>,
    pub error: Option<String>,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Eq)]
pub struct ScheduledTaskRunListRequest {
    pub page: u64,
    pub page_size: u64,
    #[serde(default)]
    pub task_code: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
}

File: apps/hook_frontend/src/types/scheduler.ts
import type { PageResponse } from './rbac';

export type ScheduledTaskConfigValueType = 'integer';

export type ScheduledTaskConfigField = {
  key: string;
  label_key: string;
  value_type: ScheduledTaskConfigValueType;
  required: boolean;
  min?: number | null;
  max?: number | null;
  unit_key?: string | null;
};

export type ScheduledTask = {
  code: string;
  name_key: string;
  description_key: string;
  enabled: boolean;
  interval_seconds: number;
  next_run_at?: string | null;
  config: Record<string, unknown>;
  config_schema: ScheduledTaskConfigField[];
  last_started_at?: string | null;
  last_finished_at?: string | null;
  last_status?: string | null;
  last_duration_ms?: number | null;
  last_error?: string | null;
  created_at: string;
  updated_at: string;
};

export type ScheduledTaskRun = {
  id: string;
  task_code: string;
  status: string;
  started_at: string;
  finished_at?: string | null;
  duration_ms?: number | null;
  message?: string | null;
  error?: string | null;
};

export type ScheduledTaskUpdate = Partial<{
  enabled: boolean;
  interval_seconds: number;
  config: Record<string, unknown>;
}>;

export type ScheduledTaskRunFilters = {
  status?: string;
  task_code?: string;
};

export type ScheduledTaskRunPage = PageResponse<ScheduledTaskRun>;
